lib/auth/keystore/keystore_test.go

```diff
@@ -0,0 +1,288 @@
+AQH/BAUwAwEB/zAdBgNVHQ4EFgQUAJprFmwUDaYguqQJxHLBC35BeQ0wDQYJKoZI
+hvcNAQELBQADggEBAG3k42nHnJvsf3M4EPBqMFqLHJOcwt5jpfPrjLmTCAnjialq
+v0qp/JAwC3SgrDFQMNNcWTi+H+E9FqYVavysZbBd0/cFlYH9SWe9CJi5CyfbsLGD
+PX8hBRDZmmhTXMMHzyHqolVAFCf5nNQyQVeQGt3fBDh++WNjmkS+886lhag/a2hh
+nskCVzvig1/6Exp06k2mMlphC25/bNB3SDeQj+dIJdej6btZs2goQZ/5Sx/iwB5c
+xrmzqBs9YMU//QIN5ZFE+7opw5v6mbeGCCk3woH46VmVwO6mHCfLha4K/K92MMdg
+JhuTMEqUaAOZBoQLn+txjl3nu9WwTThJzlY0L4w=
+-----END CERTIFICATE-----
+`)
+	testPKCS11Key = []byte(`pkcs11:{"host_id": "server2", "key_id": "00000000-0000-0000-0000-000000000000"}`)
+
+	testRawSSHKeyPair = &types.SSHKeyPair{
+		PublicKey:      testRawPublicKey,
+		PrivateKey:     testRawPrivateKey,
+		PrivateKeyType: types.PrivateKeyType_RAW,
+	}
+	testRawTLSKeyPair = &types.TLSKeyPair{
+		Cert:    testRawCert,
+		Key:     testRawPrivateKey,
+		KeyType: types.PrivateKeyType_RAW,
+	}
+	testRawJWTKeyPair = &types.JWTKeyPair{
+		PublicKey:      testRawPublicKey,
+		PrivateKey:     testRawPrivateKey,
+		PrivateKeyType: types.PrivateKeyType_RAW,
+	}
+
+	testPKCS11SSHKeyPair = &types.SSHKeyPair{
+		PublicKey:      testRawPublicKey,
+		PrivateKey:     testPKCS11Key,
+		PrivateKeyType: types.PrivateKeyType_PKCS11,
+	}
+	testPKCS11TLSKeyPair = &types.TLSKeyPair{
+		Cert:    testRawCert,
+		Key:     testPKCS11Key,
+		KeyType: types.PrivateKeyType_PKCS11,
+	}
+	testPKCS11JWTKeyPair = &types.JWTKeyPair{
+		PublicKey:      testRawPublicKey,
+		PrivateKey:     testPKCS11Key,
+		PrivateKeyType: types.PrivateKeyType_PKCS11,
+	}
+)
+
+func TestKeyStore(t *testing.T) {
+	testcases := []struct {
+		desc       string
+		rawConfig  *keystore.RawConfig
+		shouldSkip func() bool
+		setup      func(t *testing.T)
+	}{
+		{
+			desc: "raw keystore",
+			rawConfig: &keystore.RawConfig{
+				RSAKeyPairSource: native.GenerateKeyPair,
+			},
+			shouldSkip: func() bool { return false },
+		},
+	}
+
+	for _, tc := range testcases {
+		tc := tc
+		t.Run(tc.desc, func(t *testing.T) {
+			if tc.shouldSkip() {
+				return
+			}
+			t.Parallel()
+
+			if tc.setup != nil {
+				tc.setup(t)
+			}
+
+			// create the keystore
+			keyStore := keystore.NewRawKeyStore(tc.rawConfig)
+			require.NotNil(t, keyStore)
+
+			// create a key
+			key, signer, err := keyStore.GenerateRSA()
+			require.NoError(t, err)
+			require.NotNil(t, key)
+			require.NotNil(t, signer)
+
+			// delete the key when we're done with it
+			t.Cleanup(func() { require.NoError(t, keyStore.DeleteKey(key)) })
+
+			// get a signer from the key
+			signer, err = keyStore.GetSigner(key)
+			require.NoError(t, err)
+			require.NotNil(t, signer)
+
+			// try signing something
+			message := []byte("Lorem ipsum dolor sit amet...")
+			hashed := sha256.Sum256(message)
+			signature, err := signer.Sign(rand.Reader, hashed[:], crypto.SHA256)
+			require.NoError(t, err)
+			require.NotEmpty(t, signature)
+			err = rsa.VerifyPKCS1v15(signer.Public().(*rsa.PublicKey), crypto.SHA256, hashed[:], signature)
```

Comment:
Is this an important test assertion here? I.e. that if the "known-good"  implementation in `rsa` verifies the signature, then the signer implementation is good? If so, might be worth a comment.

Reply:
Yeah that's the idea, I can add a comment